pkg/route/admission/admission.go

```diff
@@ -0,0 +1,138 @@
+func NewRouteAdmissionPlugin(kClient kclient.Interface, osClient osclient.Interface) (admission.Interface, error) {
+	if kClient == nil || osClient == nil {
+		return nil, errors.New("kube client and openshift client are required")
+	}
+
+	store := cache.NewStore(StoreKeyFunc)
+	reflector := cache.NewReflector(
+		&cache.ListWatch{
+			ListFunc: func() (runtime.Object, error) {
+				return osClient.Routes(api.NamespaceAll).List(labels.Everything(), fields.Everything())
+			},
+			WatchFunc: func(resourceVersion string) (watch.Interface, error) {
+				return osClient.Routes(api.NamespaceAll).Watch(labels.Everything(), fields.Everything(), resourceVersion)
+			},
+		},
+		&routeapi.Route{},
+		store,
+		0,
+	)
+	reflector.Run()
+
+	return &routeUniqueness{
+		kClient:    kClient,
+		osClient:   osClient,
+		routeStore: store,
+	}, nil
+}
+
+// routeUniqueness implements the plugin that validates route uniqueness.
+type routeUniqueness struct {
+	kClient    kclient.Interface
+	osClient   osclient.Interface
+	routeStore cache.Store
+}
+
+// Admit ensures that the route in the form of scheme://host/path does not already exist in the
+// system.  Rules for admission:
+// 1.  If you're adding a route it must have a unique scheme, host, path combination
+// 2.  If you're updating a route you may not update it to a scheme, host, path combination that
+//     already exists (used by checking the namespace and name of the route)
+func (p *routeUniqueness) Admit(a admission.Attributes) error {
+	if a.GetResource() != "routes" {
+		return nil
+	}
+	if a.GetOperation() == "DELETE" {
+		return nil
+	}
+
+	route, ok := a.GetObject().(*routeapi.Route)
+	if !ok {
+		return apierrors.NewBadRequest("Resource was marked with kind Route but was unable to be converted")
+	}
+	routeKey := MakeRouteKey(route)
+
+	// see if we can find a route that matches the scheme, host, path combination
+	foundObj, exists, err := p.routeStore.GetByKey(routeKey)
+	if err != nil {
+		return err
+	}
+	// if no existing route is found this is ok to create
+	if !exists {
+		return nil
+	}
+	// if updating then the namespace and name must match in order to accept the update
+	found, ok := foundObj.(*routeapi.Route)
+	if !ok {
+		return fmt.Errorf("Invalid object found in store, unable to convert to route %v", foundObj)
+	}
+
+	// if adding and the route already exists then deny
+	if a.GetOperation() == "ADD" && exists {
+		return apierrors.NewForbidden("Route", route.Name, fmt.Errorf("Route %s already exists.  If you are the owner of this domain please contact support.", routeKey))
+	}
+
+	if found.Namespace != route.Namespace || found.Name != route.Name {
+		return apierrors.NewForbidden("Route", route.Name, fmt.Errorf("Route %s already exists.  If you are the owner of this domain please contact support.", routeKey))
+	}
+
+	// found a route, uids match, ok to update
+	return nil
+}
+
+func (p *routeUniqueness) Handles(operation admission.Operation) bool {
+	return true
```

Comment:
You seem to ignore DELETE entirely, so maybe update this method to just return true for admission.Create and admission.Update... you do not use CONNECT either, so yeah, just make this handler only be invoked on admission.Create and admission.Update